src/metadata/trackmetadata.cpp

```diff
@@ -0,0 +1,142 @@
+#include "metadata/trackmetadata.h"
+
+#include "util/math.h"
+
+namespace Mixxx {
+
+/*static*/ const double TrackMetadata::BPM_UNDEFINED = 0.0;
+/*static*/ const double TrackMetadata::BPM_MIN = 0.0; // lower bound (inclusive)
+/*static*/ const double TrackMetadata::BPM_MAX = 300.0; // lower bound (inclusive)
+
+/*static*/ const double TrackMetadata::REPLAYGAIN_UNDEFINED = 0.0;
+/*static*/ const double TrackMetadata::REPLAYGAIN_MIN = 0.0; // lower bound (inclusive)
+/*static*/ const double TrackMetadata::REPLAYGAIN_0DB = 1.0;
+
+double TrackMetadata::parseBpm(const QString& sBpm, bool* pValid) {
+    if (pValid) {
+        *pValid = false;
+    }
+    if (sBpm.trimmed().isEmpty()) {
+        return BPM_UNDEFINED;
+    }
+    bool bpmValid = false;
+    double bpm = sBpm.toDouble(&bpmValid);
+    if (bpmValid) {
+        if (BPM_UNDEFINED == bpm) {
+            // special case
+            if (pValid) {
+                *pValid = true;
+            }
+            return bpm;
+        }
+        while (BPM_MAX < bpm) {
+            // Some applications might store the BPM as an
+            // integer scaled by a factor of 10 or 100 to
+            // preserve fractional digits.
+            qDebug() << "Scaling BPM value:" << bpm;
+            bpm /= 10.0;
+        }
+        if (TrackMetadata::isBpmValid(bpm)) {
+            if (pValid) {
+                *pValid = true;
+            }
+            return bpm;
+        } else {
+            qDebug() << "Invalid BPM value:" << sBpm << "->" << bpm;
+        }
+    } else {
+        qDebug() << "Failed to parse BPM:" << sBpm;
+    }
+    return BPM_UNDEFINED;
+}
+
+QString TrackMetadata::formatBpm(double bpm) {
+    if (TrackMetadata::isBpmValid(bpm)) {
+        return QString::number(bpm);
+    } else {
+        return QString();
+    }
+}
+
+QString TrackMetadata::formatBpm(int bpm) {
+    if (TrackMetadata::isBpmValid(bpm)) {
+        return QString::number(bpm);
+    } else {
+        return QString();
+    }
+}
+
+namespace {
+
+const QString REPLAYGAIN_UNIT("dB");
+const QString REPLAYGAIN_SUFFIX(" " + REPLAYGAIN_UNIT);
+
+} // anonymous namespace
+
+double TrackMetadata::parseReplayGain(QString sReplayGain, bool* pValid) {
+    if (pValid) {
+        *pValid = false;
+    }
+    QString normalizedReplayGain(sReplayGain.trimmed());
```

Comment:
This comment was lost, her is it again: 

Please add the reference:
// Format is specified here:
// http://wiki.hydrogenaud.io/index.php?title=ReplayGain_1.0_specification#ID3v2